perun-contract/src/Perun/Offchain.hs

```diff
@@ -0,0 +1,572 @@
+      hash = channelHash cId
+      lookups =
+        Constraints.typedValidatorLookups (typedChannelValidator cId)
+          P.<> Constraints.plutusV2OtherScript (channelValidator cId)
+          P.<> Constraints.unspentOutputs (Map.singleton oref o)
+          P.<> (Constraints.mintingPolicy (mkVersionedChannelTokenMintinPolicy (ctTxOutRef channelToken)))
+      tx =
+        mconcat (zipWith Constraints.mustPayToPubKey (pPaymentPKs channelParameters) (map Ada.lovelaceValueOf (balances state)))
+          <> Constraints.mustValidateIn (Ledger.from $ earliestFirstCloseTime time (pTimeLock channelParameters))
+          <> Constraints.mustSpendScriptOutput oref r
+          <> (Constraints.mustMintValueWithRedeemer (Redeemer (PlutusTx.toBuiltinData (hash, Perun.Onchain.Burn))) (inv $ channelTokenValue' channelToken))
+  ledgerTx <- submitTxConstraintsWith lookups tx
+  void . awaitTxConfirmed $ getCardanoTxId ledgerTx
+  logInfo @P.String $
+    printf
+      "force closed channel %s with parameters %s. The final balances are: %s"
+      (P.show cId)
+      (P.show channelParameters)
+      (P.show $ balances state)
+
+-- | Calculates the earliest time a channel can be force-closed for given parameters.
+earliestFirstCloseTime :: POSIXTime -> Integer -> POSIXTime
+earliestFirstCloseTime timeStamp timeLock = timeStamp + POSIXTime timeLock
+
+-- | findChannelWithSync is an optimistic findChannel handler. It first tries
+-- to find a channel and if not successful, retries it again AFTER making sure
+-- to synchronize with the chainindex to its current slot.
+findChannelWithSync ::
+  (AsPerunError e, AsContractError e) =>
+  ChannelID ->
+  Contract w s e (TxOutRef, ChainIndexTxOut, ChannelDatum)
+findChannelWithSync cid =
+  handleError errorHandler $ findChannel cid
+  where
+    errorHandler (FindChannelError _) = do
+      logWarn @Text "Optimistic findChannel not succeeded. Trying again after syncing PAB with Chainindex."
+      retryAfterSync $ findChannel cid
+    errorHandler err = throwing _PerunError err
+
+-- | findChannelWithSync' is an optimistic findChannel handler. It first tries
+-- to find a valid channel for the given channel id and thread token and if not successful,
+-- retries it again AFTER making sure to synchronize with the chainindex to its current slot.
+findChannelWithSync' ::
+  (AsPerunError e, AsContractError e) =>
+  ChannelID ->
+  AssetClass ->
+  Contract w s e (TxOutRef, ChainIndexTxOut, ChannelDatum)
+findChannelWithSync' cid ct =
+  handleError errorHandler $ findChannel cid
+  where
+    errorHandler (FindChannelError _) = do
+      logWarn @Text "Optimistic findChannel not succeeded. Trying again after syncing PAB with Chainindex."
+      retryAfterSync $ findChannel' cid ct
+    errorHandler err = throwing _PerunError err
+
+getChannelId :: Channel -> ChannelID
+getChannelId channel = ChannelID . S.dataHash $ PlutusTx.toBuiltinData channel
+
+findChannel ::
```

Comment:
```suggestion
-- | Deprecated. Use findChannel' instead
findChannel ::
```

Reply:
I removed that altogether, because it is not necessary to keep for the real contract.
Our `EvilContract` had a dependency on this old impl, so I just moved the implementation to the test package.